### PR TITLE
Hadamard product

### PR DESCRIPTION
**Pull request by dominikkiese**
Implemented Hadamard (elementwise) product for nda::arrays, std::array, std::vector and built-in arithmetic types (based on suggestions by @Thoemi09 and @parcollet).

### CODE REVIEW

File: c++/nda/algorithms.hpp [hunk 1]
```diff
@@ -199,6 +200,73 @@ namespace nda {
...
+   * @brief Hadamard product of two std::vector objects.
+   * 
+   * @tparam T Data type of the first input vector.
+   * @tparam U Data type of the second input vector.
+   * @param a std::vector object.
+   * @param b std::vector object.
+   * @return std::vector containing the elementwise product of the two input vectors.
+   */
+  template <typename T, typename U>
+  [[nodiscard]] constexpr auto hadamard(std::vector<T> const &a, std::vector<U> const &b) {
```
Comment (Wentzell) [inline c++/nda/algorithms.hpp:247]: I would expect a similar behavior between std::vector and std::array. Can you generalize the array case to allow `TU` type propagation in the same way? I.e. updating the `operator*`

File: c++/nda/algorithms.hpp [hunk 2]
```diff
@@ -199,6 +200,73 @@ namespace nda {
...
+   * @param b Second input.
+   * @return Product of the two inputs.
+   */
+  template <typename T, typename U>
+  constexpr auto hadamard(T a, U b)
+    requires((std::is_arithmetic_v<T> || std::is_same_v<T, std::complex<typename T::value_type>>)
+             && (std::is_arithmetic_v<U> || std::is_same_v<U, std::complex<typename U::value_type>>))
+  {
+    return a * b;
+  }
```
Comment (Wentzell) [inline c++/nda/algorithms.hpp]:
```suggestion
  constexpr auto hadamard(nda::Scalar auto a, nda::Scalar auto b) { return a * b; }
```

File: c++/nda/algorithms.hpp [hunk 3]
```diff
@@ -199,6 +200,73 @@ namespace nda {
...
+   * @tparam A nda::Array type.
+   * @tparam B nda::Array type.
+   * @param a nda::Array object.
+   * @param b nda::Array object.
+   * @return nda::Array containing the elementwise product of the two input arrays.
+   */
+  template <Array A, Array B>
+    requires(nda::get_rank<A> == nda::get_rank<B>)
+  [[nodiscard]] constexpr auto hadamard(A &&a, B &&b) {
+    assert(a.shape() == b.shape());
```
Comment (Wentzell) [inline c++/nda/algorithms.hpp]:
```suggestion
```
Already taken care of by map

File: c++/nda/algorithms.hpp [hunk 4]
```diff
@@ -199,6 +200,73 @@ namespace nda {
...
+   * 
+   * @tparam T Data type of the arrays.
+   * @tparam R Size of the arrays.
+   * @param a std::array object.
+   * @param b std::array object.
+   * @return std::array containing the elementwise product of the two input arrays.
+   */
+  template <typename T, size_t R>
+  [[nodiscard]] constexpr auto hadamard(std::array<T, R> const &a, std::array<T, R> const &b) {
+    assert(a.size() == b.size());
```
Comment (Wentzell) [inline c++/nda/algorithms.hpp]:
```suggestion
```
Redundant, same as `R == R`

File: c++/nda/algorithms.hpp [hunk 5]
```diff
@@ -199,6 +200,73 @@ namespace nda {
...
+   * @tparam T Data type of the first input vector.
+   * @tparam U Data type of the second input vector.
+   * @param a std::vector object.
+   * @param b std::vector object.
+   * @return std::vector containing the elementwise product of the two input vectors.
+   */
+  template <typename T, typename U>
+  [[nodiscard]] constexpr auto hadamard(std::vector<T> const &a, std::vector<U> const &b) {
+    using TU = decltype(std::declval<T>() * std::declval<U>());
+    assert(a.size() == b.size());
```
Comment (Wentzell) [inline c++/nda/algorithms.hpp]:
```suggestion
    EXPECTS(a.size() == b.size());
```

**Review by Wentzell — CHANGES_REQUESTED** [review]
Thank you @dominikkiese for this new feature! See below for suggestions

File: c++/nda/algorithms.hpp (hunk 1 above)
Comment (dominikkiese) [reply c++/nda/algorithms.hpp:247]: Checks out! Should be fixed by b3df66ac48ba3d045d2c5b9ed871e319388bbe05. What do you think?

**Review by Wentzell — APPROVED** [review]
Thank you @dominikkiese, Looks good now!

**Review by Thoemi09 — APPROVED** [review]

**Pull request merged**